[android] Enable navigation in background mode

rtsisyk opened the pull request:
TODO:

- [x] Synchronize LocationHelper and NavigationService
- [x] Figure out how to avoid ForegroundServiceStartNotAllowedException on API 31+
- [x] Add dialog to explain why permissions are needed - need texts, show the dialog before asking for POST_NOTIFICATIONS
- [x] Don't disable TTS in onPause()
- [x] Improve UI of notification in the status bar

Closes #573

arnaudvergnet commented:
Will this mean OM will stop loosing the GPS position when locking the screen, or is it only when using the navigation feature?

shankarpriyank commented on android/AndroidManifest.xml:
```diff
@@ -25,6 +25,7 @@
   <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
   <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
   <uses-permission android:name="android.permission.ACCESS_LOCATION_EXTRA_COMMANDS" />
+  <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
```
Is this really necessary to ask for this permission?
This can really complicate things

rtsisyk replied:
Yes, ACCESS_BACKGROUND_LOCATION permission is needed for the background location ;)

AndrewShkrob commented:
I'd suggest moving instance of NavigationService from NavigationController to MwmApplication. This will help me a lot during Android Auto development.

rtsisyk commented on android/src/app/organicmaps/MwmActivity.java:
```diff
@@ -1479,6 +1498,44 @@ public void onNavigationStarted()
+        new MaterialAlertDialogBuilder(this, R.style.MwmTheme_AlertDialog)
+            // TODO: text
+            .setTitle("Background location")
+            .setMessage("bla bla")
```
TODO: add information about background location

rtsisyk commented on android/src/app/organicmaps/routing/NavigationService.java:
```diff
@@ -215,52 +177,33 @@ private Notification getNotification()
+    if (!routingController.isNavigating())
+      return;
+
     final String[] turnNotifications = Framework.nativeGenerateNotifications();
```
TODO: review the design of the notification

rtsisyk commented:
https://github.com/organicmaps/organicmaps/assets/1799054/21434999-e804-454f-9278-4fc47dee1ede

https://github.com/organicmaps/organicmaps/assets/1799054/fbab47be-1a75-4503-a011-6937d61d0c25

rtsisyk commented:
https://github.com/organicmaps/organicmaps/assets/1799054/dc39482f-0306-41d4-8ab6-1a85faca4c76

https://github.com/organicmaps/organicmaps/assets/1799054/d28b9b1e-6d33-4ba3-bc70-ea72a6f12f65

biodranik commented:
Should background mode always activate, or should it be configurable, to enable the current behavior (for example, to save the battery)?

rtsisyk commented:
> Should background mode always activate, or should it be configurable, to enable the current behavior (for example, to save the battery)?

In the current version, the background mode is activated during navigation. I don't see any use-cases for navigation without location.

biodranik commented:
The current production version navigates without background location and saves a lot of battery. Users need to wait a bit on each app activation, but it may worth the battery saving.

Jean-BaptisteC reviewed (changes requested):
❌ Pixel 6 - Android 13

- Stacktrace appears when I rotate device when pop-up to enable background location is on screen (when i rotate device, pop-up disappear)
```
android.view.WindowLeaked: Activity app.organicmaps.MwmActivity has leaked window DecorView@ce577f7[MwmActivity] that was originally added here
at android.view.ViewRootImpl.<init>(ViewRootImpl.java:943)
at android.view.ViewRootImpl.<init>(ViewRootImpl.java:931)
at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:391)
at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:148)
at android.app.Dialog.show(Dialog.java:352)
at androidx.appcompat.app.AlertDialog$Builder.show(AlertDialog.java:1009)
at app.organicmaps.MwmActivity.onNavigationStarted(MwmActivity.java:1525)
at app.organicmaps.routing.RoutingController.startNavigation(RoutingController.java:577)
at app.organicmaps.routing.RoutingController.start(RoutingController.java:430)
at app.organicmaps.MwmActivity.onRoutingStart(MwmActivity.java:1727)
at app.organicmaps.routing.RoutingBottomMenuController.lambda$setStartButton$0$app-organicmaps-routing-RoutingBottomMenuController(RoutingBottomMenuController.java:214)
at app.organicmaps.routing.RoutingBottomMenuController$$ExternalSyntheticLambda0.onClick(Unknown Source:2)
at android.view.View.performClick(View.java:7542)
at android.view.View.performClickInternal(View.java:7519)
at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
at android.view.View$PerformClick.run(View.java:29476)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7918)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

When I want open OM from notifications routing is canceled (I didn't touch exit button in notification, I touche entire panel of notifications) 

Last thing, arrow is not easy to see in Android 13, with [dynamic color](https://developer.android.com/develop/ui/views/theming/dynamic-colors)
<img src="https://github.com/organicmaps/organicmaps/assets/87148630/1db33a0d-aba9-469f-8ac1-df6b0a00d12a" height=400 />

rtsisyk commented:
> The current production version navigates without background location and saves a lot of battery. Users need to wait a bit on each app activation, but it may worth the battery saving.

Don't grant background location permission and will not be used.

rtsisyk commented:
@Jean-BaptisteC , thanks for testing. I can confirm all three issues:

>  Stacktrace appears when I rotate device when pop-up to enable background location is on screen (when i rotate device, pop-up disappear)

> When I want open OM from notifications routing is canceled (I didn't touch exit button in notification, I touche entire panel of notifications)

> Last thing, arrow is not easy to see in Android 13, with [dynamic color](https://developer.android.com/develop/ui/views/theming/dynamic-colors)

biodranik commented:
> > The current production version navigates without background location and saves a lot of battery. Users need to wait a bit on each app activation, but it may worth the battery saving.
> 
> 
> 
> Don't grant background location permission and will not be used.
> 
> 

1. Changing the important feature's behavior should be intuitive and discoverable by users. Likely in settings or maybe also in the navigation panel. 
2. The most important UX question: when background permissions should be requested and re-requested?

rtsisyk commented:
I want to rebase this patch on the top of #4602

Jean-BaptisteC commented:
Another thing when I leave navigation from notifications -> I go back to app
Notifications are never disappeared (notification stay with older navigation instructions), to remove notifications, I need to force stop

rtsisyk commented:
> Stacktrace appears when I rotate device when pop-up to enable background location is on screen (when i rotate device, pop-up disappear)

Fixed.

> When I want open OM from notifications routing is canceled (I didn't touch exit button in notification, I touche entire panel of notifications)

Fixed.

> Last thing, arrow is not easy to see in Android 13, with [dynamic color](https://developer.android.com/develop/ui/views/theming/dynamic-colors)

Fixed, I reworked UI completely (see video below).

> Another thing when I leave navigation from notifications -> I go back to app
> Notifications are never disappeared (notification stay with older navigation instructions), to remove notifications, I need to force stop

Fixed.

New reworked notification:

https://github.com/organicmaps/organicmaps/assets/1799054/615a3688-be55-402c-89fc-3d6224762ba6

rtsisyk commented on android/src/app/organicmaps/routing/NavigationService.java:
```diff
@@ -1,110 +1,128 @@
 package app.organicmaps.routing;
```
This file was 100% rewritten. Don't try to read the diff.

rtsisyk commented on android/src/app/organicmaps/MwmActivity.java:
```diff
@@ -1736,6 +1750,62 @@ private void startLocation()
+
+      dismissLocationErrorDialog();
+      mLocationErrorDialog = new MaterialAlertDialogBuilder(this, R.style.MwmTheme_AlertDialog)
+          // TODO: add disclaimer text
```
TODO: add proper text.

Jean-BaptisteC commented:
On new version, i can't accept or deny background location pop-up, because this pop-up blink and disappear

rtsisyk commented:
> On new version, I can't accept or deny background location pop-up, because this pop-up blinks and disappear
> 

Fixed. Yes, of course, this dialog must not disappear when location is received. All other location dialogs are automatically dismissed when location is received.

rtsisyk commented:
I don't have any good solution for AlertDialogs dismissed during screen rotation (onPause of MwmActivity). It looks like that Android handles onPause()/onResume() properly for Fragments, but not for Activites. We should try [DialogFragment](https://developer.android.com/reference/android/app/DialogFragment) instead of bare AlertDialog. Let's address this issue separately to avoid side tracking in this PR. MwmActivity alon has 10 such dialogs, this new disclaimer about background location will 11th.

rtsisyk commented:
This PR is basically code completed now. I plan to test older API levels and add a proper message to the dialog.

Jean-BaptisteC commented:
I think we could disable navigation in background when Power/Battery Saver mode is enabled?
Other thing, when you near at roundabout, exit number is not present in notifications, but I think implementing this needs some changes C++ code, right?

rtsisyk commented:
The only one issue I see is that on emualtor with API=33 the alert dialog disappears quickly because MwmActivtiy does onPause() without any clear reason. On the real device this issue doesn't repro.  It could be an emulator-specific issue. I what causes onPause()

rtsisyk commented:
The issue with disappearing disclaimer about background location has been fixed. Kudos to @AndrewShkrob for finding the bug.

botanegg commented:
Telegram discussion here: https://t.me/OrganicMapsRu/40307

rtsisyk commented:
Added the restart of location from onStartCommand(), as suggested by @botanegg.

Jean-BaptisteC reviewed (approved):
✅ Pixel 6 - Android 13

biodranik commented on android/src/app/organicmaps/MwmActivity.java:
```diff
@@ -855,8 +856,6 @@ protected void onSaveInstanceState(@NonNull Bundle outState)
         fragment.saveRoutingPanelState(outState);
     }
 
-    mNavigationController.onActivitySaveInstanceState(this, outState);
```
What happens if the app is killed by the system? AFAIR the current implementation restored and continued navigation.

biodranik replied:
@Jean-BaptisteC can you please test this case on the current master and on this branch?

rtsisyk replied:
mNavigationController.onActivitySaveInstanceState() here was no-op, navigation state is saved in a different place. I tested that navigation is restarted when the app was killed.

biodranik commented on android/src/app/organicmaps/MwmActivity.java:
```diff
@@ -1539,9 +1533,18 @@ public void onNavigationStarted()
     mMapButtonsViewModel.setLayoutMode(MapButtonsController.LayoutMode.navigation);
     refreshLightStatusBar();
+
+    // Don't start the background navigation service without fine location.
```
Does it make sense to start foreground navigation without the fine location?

rtsisyk replied:
No, it doesn't make sense. COARSE location returns some random point in the city. It is not usable for voice instructions & navigation.

biodranik replied:
This comment implies that navigation can be started with a coarse location. Does it?

rtsisyk replied:
Yes, in the current master "START" button can be pressed even without location. This patch doesn't change this behavior. The service is not started without FINE location because it is not really useful.

biodranik commented on data/strings/strings.txt:
```diff
@@ -21785,6 +21743,48 @@
     zh-Hans = 加载地图
     zh-Hant = 載入地圖
 
+  [notification_channel_navigation]
```
It already exists in this file:
```
  [prefs_group_route]
    tags = android,ios
    en = Navigation
    af = Navigasie
    ar = الملاحة
    be = Навігацыя
    bg = Навигация
```

biodranik replied:
There's no need to duplicate the translation if the meaning is _exactly_ the same. You can rename the key to [navigation] and reuse it where necessary.

rtsisyk replied:
I will re-use existing string, no problem.

rtsisyk commented:
Rebased on the top of new Android structure.
